### PR TITLE
Fix the bug that `validate` flag is not set when the SAN(SubjectAltName) matching is performed

### PR DESCRIPTION
**Pull request by tyxia**
When SAN(SubjectAltName) matching with the provided matchers is the only verification performed and matching succeeds, the `validated` flag is not updated properly(i.e. left as `NotValidated`) and the function return wrong status code.

Add the unit test:
1. Create `DefaultValidator` object with NULL certificateValidationContext (i.e no validation context). This ensures that the certificate verification later will be only performed on the SAN matcher arg provided in step#3
2. Create the cert from test data file.
3. Provide the SAN matcher as only input and perform the certificate verification with correct SAN regex matcher and incorrect SAN exact matcher(negative test).

Risk Level: Low
Testing:
Docs Changes: n/a
Release Notes: n/a

Signed-off-by: Tianyu Xia <tyxia@google.com>

### CODE REVIEW

**Review by rojkov — COMMENTED** [review]
Thank you for fix it! Could you please add a test case for it?

**Comment by tyxia** [comment]
> Thank you for fix it! Could you please add a test case for it?

Thank you for the review!
This code path is currently not tested in default_validator_test unit test. In order to test this path quickly, I coded up a small test to : 1) set up TLS ClientContext 2) load the certificate yaml test data (with customization for SAN matcher) 3) call the function with SSL-specific structs. 
The code works as expected, but I found an interesting thing: 
The caller site (around line#225 in this file, which normally will be called from boringssl) **only cares if the validate flag is Envoy::Ssl::ClientValidationStatus::Failed or not.**

I think it is because:
1. This function(i.e. function `DefaultCertValidator::verifyCertificate`) is used to verify context configuration like subject name, hash etc, but not for trust_ca.
2. `only trust_ca is specified but no subject name, hash, etc` is a valid validation context configuration. This matches the code in `DefaultCertValidator::initializeSslContexts` and  document [here](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl)
3. And as we can see in doVerifyCertChain function, trust_ca path is validated separately.

So, **the validate flag in this function I changed only represents that it is not validated for subject name, hash etc but doesn't mean that it is not validated because we could have only trust_ca case that is validated in the code above.** 
 
In short, I think the fix is technically correct but doesn't make the difference in practice based on current code. Also, this code path is already tested in many integration tests(not its unit test though). Thus, we may not need a new test case here.

Please let me know what do you think.

**Comment by rojkov** [comment]
Yes, I have no doubt the code is correct. But a unit test could help with making it future-proof and to pinpoint a problem better.

We know that in case of `DefaultCertValidator` its method `verifyCertificate()` should return either Validated or Failed status. So for the sake of future refactorings it'd be beneficial to check this invariant in a unit test.

For example the function could be simplified even farther if we dropped `validated` completely and returned `Envoy::Ssl::ClientValidationStatus::Validated` in the last `return`.

**Comment by tyxia** [comment]
> Yes, I have no doubt the code is correct. But a unit test could help with making it future-proof and to pinpoint a problem better.
> 
> We know that in case of `DefaultCertValidator` its method `verifyCertificate()` should return either Validated or Failed status. So for the sake of future refactorings it'd be beneficial to check this invariant in a unit test.
> 
> For example the function could be simplified even farther if we dropped `validated` completely and returned `Envoy::Ssl::ClientValidationStatus::Validated` in the last `return`.

Thanks for

**Comment by tyxia** [comment]
> Yes, I have no doubt the code is correct. But a unit test could help with making it future-proof and to pinpoint a problem better.
> 
> We know that in case of `DefaultCertValidator` its method `verifyCertificate()` should return either Validated or Failed status. So for the sake of future refactorings it'd be beneficial to check this invariant in a unit test.
> 
> For example the function could be simplified even farther if we dropped `validated` completely and returned `Envoy::Ssl::ClientValidationStatus::Validated` in the last `return`.

**Comment by tyxia** [comment]
> > Yes, I have no doubt the code is correct. But a unit test could help with making it future-proof and to pinpoint a problem better.
> > We know that in case of `DefaultCertValidator` its method `verifyCertificate()` should return either Validated or Failed status. So for the sake of future refactorings it'd be beneficial to check this invariant in a unit test.
> > For example the function could be simplified even farther if we dropped `validated` completely and returned `Envoy::Ssl::ClientValidationStatus::Validated` in the last `return`.

Thank you for inputs! I have some local test code but I have one thing want to discuss first:
**Should we treat `Envoy::Ssl::ClientValidationStatus::NotValidated` as a `OK` return code at caller site (line#225 in this file)?** (i.e. always ignore `NotValidated` and return `Validated` in the last `return`).

Without full background info of the code here but based on my code reading, I think there is one corner case not being handled right now : `trusted_ca = false` and `no hash or SAN matcher` . 
Basically, it means no validation context is provided so no validation is performed. Current code will still return `OK` for this case but technically it seems not correct that we `verifyCertificate` callback get called but no certificate validation_context provided?

I quickly tested one change at line 225 to cover this case
from:
```
return allow_untrusted_certificate_ ? 1
                                      : (validated != Envoy::Ssl::ClientValidationStatus::Failed);
```
to
```
return allow_untrusted_certificate_ ? 1 : verify_trusted_ca_ ? validated != Envoy::Ssl::ClientValidationStatus::Failed : validated == Envoy::Ssl::ClientValidationStatus::Validated;
```
And it seems reveal one potential bug in sds_dynamic_integration_test's QUIC path(The only failure caused by change above). Basically, the problem is QUIC register and then will invoke the  `verifyCertificate()` callback by setting cert_verify_mode to `SSL_VERIFY_PEER`. But the test doesn't provide the validation_context. Thus, it will run into `NotValidated` case and current code basically ignore it.

The quick experimental fix for that could be query the true verify_mode that is initialized in `DefaultCertValidator::initializeSslContexts`, it will be `SSL_VERIFY_NONE` for this case because there is no validation_context config provided. 
And then similar to [openSSL logic](https://github.com/google/boringssl/blob/e9c5d72c09e01a0f71f30f7c3454e5e7f8711476/ssl/ssl_x509.cc#L409), we can treat the the error as non-fatal if mode is SSL_VERIFY_NONE.
I have a draft PR (not the final) #16982 as a reference to what are changes I am making.

I feel this might not be a critical issue/bug :) but it also doesn't hurt to get to the bottom of it. 
Thanks for your time and review!

**Comment by rojkov** [comment]
This sounds reasonable. To me `NotValidated` feels like an initial state that needs to be driven to either `Failed` or `Validated` eventually. Consequently treating it as OK in functions like `doVerifyCertChain()` is rather wrong. But that fix can be done in a follow-up PR to keep this one small.

**Comment by tyxia** [comment]
> This sounds reasonable. To me `NotValidated` feels like an initial state that needs to be driven to either `Failed` or `Validated` eventually. Consequently treating it as OK in functions like `doVerifyCertChain()` is rather wrong. But that fix can be done in a follow-up PR to keep this one small.

Yes, I agreed. I will split those fixes into another PR.

The unit test is added. PTAL!

File: test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc [hunk 1]
```diff
@@ -116,6 +116,34 @@ TEST(DefaultCertValidatorTest, TestMatchSubjectAltNameNotMatched) {
...
+
+  bssl::UniquePtr<X509> cert = readCertFromFile(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns_cert.pem"));
+  envoy::type::matcher::v3::StringMatcher matcher;
+  matcher.MergeFrom(TestUtility::createRegexMatcher(".*.example.com"));
+  std::vector<Matchers::StringMatcherImpl> san_matchers;
+  san_matchers.push_back(Matchers::StringMatcherImpl(matcher));
+  // Verify the certificate with correct SAN regex matcher.
+  EXPECT_EQ(default_validator->verifyCertificate(cert.get(), /*verify_san_list=*/{}, san_matchers),
+            Envoy::Ssl::ClientValidationStatus::Validated);
```
Comment (rojkov) [inline test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc:136]: Also correctness of `stats` could be checked here: no increment in case of a matching SAN and incremented otherwise.

**Review by rojkov — COMMENTED** [review]
Thank you! LGTM modulo one more check.

File: test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc (hunk 1 above)
Comment (tyxia) [reply test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc:136]: Thanks for review!   I explored a few options for testing `stats`:
1. test_server_->counter(....). The problem is this test is free function test (i.e. no test fixture) so it doesn't have necessary infra setup. I could try to add it though but I feel it is a bit overkill here.
2. Exposing `stats_` in `DefaultCertValidator` class. But I feel it is not ideal to do this just for testing purpose.

I am new to Envoy codebase and I am wondering if you have other suggested ways to check `stats` easily? Thanks

**Comment by tyxia** [comment]
Also, cc @htuch for visibility.

File: test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc (hunk 1 above)
Comment (rojkov) [reply test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc:136]: No need to expose anything. `stats` are created in the line 121 and accessible farther down the test. Just put something like
```c++
EXPECT_EQ(0, stats.fail_verify_san_.value());
```
where SAN is expected to be valid and
```c++
EXPECT_EQ(1, stats.fail_verify_san_.value());
```
otherwise.

Comment (tyxia) [reply test/extensions/transport_sockets/tls/cert_validator/default_validator_test.cc:136]: Done. thanks! I didn't notice that it is pass-by-ref and reference class member. I guess this is envoy's intended dependency pattern and disallow copying (i.e. copy assignment) and moving. There should be no lifetime concern here.

**Review by rojkov — APPROVED** [review]
Thank you! Looks good to me.
@yanavlasov I think it's ready for a second pass.

**Review by yanavlasov — APPROVED** [review]
